In [29]:
import pandas as pd
import ast

We will first read the csv file titled RAW_recipes.

In [30]:
RAW_recipes=pd.read_csv('RAW_recipes.csv')

In [31]:
RAW_recipes.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2                   all in the kitchen  chili  112140      130   
3                          alouette  potatoes   59389       45   
4          amish  tomato ketchup  for canning   44061      190   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2          196586  2005-02-25   
3           68585  2003-04-14   
4           41706  2002-10-25   

                                                tags  \
0  ['60-minutes-or-less', 'time-to-make', 'course...   
1  ['30-minutes-or-less', 'time-to-make', 'course...   
2  ['time-to-make', 'course', 'preparation', 'mai...   
3  ['60-minutes-or-less', 'time-to-make', 'course...   
4  ['weeknight', 'time-to-make', 'course', 'main-...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['brown ground beef in large pot', 'add choppe...   
3  ['place potatoes in a large pot of lightly sal...   
4  ['mix all ingredients& boil for 2 1 / 2 hours ...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...              7  
1  ['prepared pizza crust', 'sausage patty', 'egg...              6  
2  ['ground beef', 'yellow onions', 'diced tomato...             13  
3  ['spreadable cheese with garlic and herbs', 'n...             11  
4  ['tomato juice', 'apple cider vinegar', 'sugar...              8

In [32]:
RAW_recipes['tags'] = RAW_recipes['tags'].apply(ast.literal_eval)
RAW_recipes['steps'] = RAW_recipes['steps'].apply(ast.literal_eval)
RAW_recipes['ingredients'] = RAW_recipes['ingredients'].apply(ast.literal_eval)

In [33]:
#!pip uninstall inflect typeguard typing-extensions -y
#!pip install typing-extensions==4.7.1
#!pip install inflect==5.3.0

In [34]:
import inflect
# Initialize the inflect engine
p = inflect.engine()

In [35]:
from itertools import chain
all_ingredients_list = list(set(chain.from_iterable(RAW_recipes['ingredients'])))

In [36]:
A=len(all_ingredients_list)

In [37]:
# Define a function to convert individual ingredient strings to singular using inflect
def to_singular(ingredient):
    return p.singular_noun(ingredient) or ingredient

# Define a function to process a list of ingredients
def process_ingredient_list(ingredient_list):
    return [to_singular(ingredient) for ingredient in ingredient_list]

In [38]:
# Apply the function to the ingredients column
RAW_recipes['ingredients'] = RAW_recipes['ingredients'].apply(process_ingredient_list)

In [39]:
all_ingredients_list = list(set(chain.from_iterable(RAW_recipes['ingredients'])))

In [40]:
B=len(all_ingredients_list)

In [41]:
C=A-B
print(C)

1210


In [42]:
# Save the list of ingredients to a text file
with open('all_ingredients_list.txt', 'w') as f:
    for ingredient in all_ingredients_list:
        f.write(f"{ingredient}\n")

We will write a function that will help us to read the "tags" column of RAW_recipes and look for certain cuisine related keywords. The following is keyword for the 100 best cuisines in the world given by Taste Atlas.

In [14]:
world_cuisines = [
    'italian',
    'japanese',
    'greek',
    'portuguese',
    'chinese',
    'indonesian',
    'mexican',
    'french',
    'spanish',
    'peruvian',
    'indian',
    'brazilian',
    'polish',
    'argentinian',
    'turkish',
    'states',
    'thai',
    'korean',
    'croatian',
    'serbian',
    'hungarian',
    'vietnamese',
    'iranian',
    'chilean',
    'lebanese',
    'georgian',
    'bosnian',
    'colombian',
    'romanian',
    'bulgarian',
    'malaysian',
    'german',
    'filipino',
    'belgian',
    'czech',
    'austrian',
    'swiss',
    'lithuanian',
    'english',
    'algerian',
    'russian',
    'slovak',
    'canadian',
    'swedish',
    'dutch',
    'moroccan',
    'scottish',
    'ecuadorian',
    'danish',
    'australian',
    'egyptian',
    'south-african',
    'ukrainian',
    'syrian',
    'irish',
    'singaporean',
    'pakistani',
    'puerto',
    'norwegian',
    'bolivian',
    'macedonian',
    'israeli',
    'palestinian',
    'slovenian',
    'finnish',
    'tunisian',
    'haitian',
    'jamaican',
    'armenian',
    'venezuelan',
    'belarusian',
    'moldovan',
    'lankan',
    'jordanian',
    'cuban',
    'uzbekistani',
    'azerbaijani',
    'taiwanese',
    'uruguayan',
    'montenegrin',
    'ethiopian',
    'iraqi',
    'qatari',
    'trinidadian',
    'libyan',
    'lao',
    'barbadian',
    'cypriot',
    'bengali',
    'kazakhstani',
    'albanian',
    'kyrgyzstani',
    'burmese',
    'zealand',
    'saudi',
    'irish',
    'bahamian',
    'dominican',
    'welsh',
    'ghanaian'
]

We will write a function that searches entries of "tags" column of the RAW_recipes for the keywords above. If there is a string which has the keyowrds above we return them.

In [15]:
def extract_cuisine_tags(tags):
    cuisine_related_tags = []
    for tag in tags:
        for cuisine in world_cuisines:
            if cuisine in tag:
                cuisine_related_tags.append(tag)
    return cuisine_related_tags

# Create a new column in the RAW_recipes with the extracted cuisine-related tags
RAW_recipes['Cuisine_Tags'] = RAW_recipes['tags'].apply(extract_cuisine_tags)

In [16]:
RAW_recipes.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2                   all in the kitchen  chili  112140      130   
3                          alouette  potatoes   59389       45   
4          amish  tomato ketchup  for canning   44061      190   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2          196586  2005-02-25   
3           68585  2003-04-14   
4           41706  2002-10-25   

                                                tags  \
0  [60-minutes-or-less, time-to-make, course, mai...   
1  [30-minutes-or-less, time-to-make, course, mai...   
2  [time-to-make, course, preparation, main-dish,...   
3  [60-minutes-or-less, time-to-make, course, mai...   
4  [weeknight, time-to-make, course, main-ingredi...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  [make a choice and proceed with recipe, depend...   
1  [preheat oven to 425 degrees f, press dough in...   
2  [brown ground beef in large pot, add chopped o...   
3  [place potatoes in a large pot of lightly salt...   
4  [mix all ingredients& boil for 2 1 / 2 hours ,...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  \
0  [winter squash, mexican seasoning, mixed spice...              7   
1  [prepared pizza crust, sausage patty, egg, mil...              6   
2  [ground beef, yellow onion, diced tomato, toma...             13   
3  [spreadable cheese with garlic and herbs, new ...             11   
4  [tomato juice, apple cider vinegar, sugar, sal...              8   

                   Cuisine_Tags  
0                     [mexican]  
1  [northeastern-united-states]  
2                            []  
3                            []  
4  [northeastern-united-states]

Now we will try to understand how many recipies are there with a Cuisine Tag we created.

In [17]:
# Count the occurrences of each category in the Cuisine_Tags column
cuisine_category_counts = RAW_recipes['Cuisine_Tags'].value_counts()

# Print the counts of all categories for reference
print(cuisine_category_counts)

[]                                                176104
[italian]                                           6911
[southern-united-states]                            6277
[mexican]                                           5642
[canadian]                                          3923
                                                   ...  
[russian, czech]                                       1
[malaysian, chinese, vietnamese]                       1
[southwestern-united-states, moroccan]                 1
[brazilian, mexican, peruvian, venezuelan]             1
[canadian, german, northeastern-united-states]         1
Name: Cuisine_Tags, Length: 667, dtype: int64


For the training purposes, we will clean the ones without a Cuisine Tag and the ones with more than a Cuisine Tag.

In [18]:
# Get the indices of rows that do not have exactly one cuisine-related tag
indices_no_cuisine_tags = RAW_recipes[RAW_recipes['Cuisine_Tags'].apply(lambda x: len(x) != 1)].index

# Drop them
RAW_recipes_with_one_cuisine = RAW_recipes.drop(indices_no_cuisine_tags).reset_index(drop=True)

In [19]:
# Count the occurrences of each category in the Cuisine_Tags
cusine_category_counts = RAW_recipes_with_one_cuisine['Cuisine_Tags'].value_counts()

# Print the counts of all categories for reference
print(cusine_category_counts)

[italian]                       6911
[southern-united-states]        6277
[mexican]                       5642
[canadian]                      3923
[indian]                        2564
[australian]                    2355
[southwestern-united-states]    2194
[greek]                         2165
[chinese]                       1841
[french]                        1813
[northeastern-united-states]    1724
[german]                        1186
[english]                       1173
[thai]                          1116
[spanish]                        871
[moroccan]                       839
[japanese]                       768
[new-zealand]                    405
[scottish]                       381
[swedish]                        340
[russian]                        322
[vietnamese]                     314
[south-african]                  313
[swiss]                          302
[polish]                         299
[portuguese]                     292
[hungarian]                      263
[

There are 6 recipes with a tag "chinese-new-year". We will turn them into just "chinese".

In [20]:
# Define a function to replace 'chinese-new-year' with 'chinese'
def replace_chinese_new_year(tags):
    return ['chinese' if tag == 'chinese-new-year' else tag for tag in tags]

# Apply the function to the 'Cuisine_Tags' column
RAW_recipes_with_one_cuisine['Cuisine_Tags'] = RAW_recipes_with_one_cuisine['Cuisine_Tags'].apply(replace_chinese_new_year)

Now we will check if everything went as we wanted.

In [21]:
# Count the occurrences of each category in the Time_Category column
cusine_category_counts = RAW_recipes_with_one_cuisine['Cuisine_Tags'].value_counts()

# Print the counts of all categories for reference
print(cusine_category_counts)

[italian]                       6911
[southern-united-states]        6277
[mexican]                       5642
[canadian]                      3923
[indian]                        2564
[australian]                    2355
[southwestern-united-states]    2194
[greek]                         2165
[chinese]                       1847
[french]                        1813
[northeastern-united-states]    1724
[german]                        1186
[english]                       1173
[thai]                          1116
[spanish]                        871
[moroccan]                       839
[japanese]                       768
[new-zealand]                    405
[scottish]                       381
[swedish]                        340
[russian]                        322
[vietnamese]                     314
[south-african]                  313
[swiss]                          302
[polish]                         299
[portuguese]                     292
[hungarian]                      263
[

In [22]:
len(RAW_recipes_with_one_cuisine)

50399

In [23]:
len(cusine_category_counts)

58

So, we have categorized 50399 recipes into 58 cuisine categories. We can use this data set as a training set for a model that predicts the cuisine for a given list of ingredients or a recipe.

In [24]:
# Write the DataFrame to a CSV file
RAW_recipes_with_one_cuisine.to_csv('RAW_recipes_with_one_cuisine.csv', index=False)

# Verify the file creation by printing a message
print("RAW_recipes_with_one_cuisine DataFrame has been written to RAW_recipes_with_one_cuisine.csv")

RAW_recipes_with_one_cuisine DataFrame has been written to RAW_recipes_with_one_cuisine.csv


In [25]:
# Get the indices of rows that do not have exactly one cuisine-related tag
indices_different_than_one_cuisine_tag= RAW_recipes[RAW_recipes['Cuisine_Tags'].apply(lambda x: len(x)==1)].index

# Drop them
RAW_recipes_without_one_cuisine = RAW_recipes.drop(indices_different_than_one_cuisine_tag).reset_index(drop=True)

In [26]:
len(RAW_recipes_without_one_cuisine)+len(RAW_recipes_with_one_cuisine)==len(RAW_recipes)

True

In [27]:
# Write the DataFrame to a CSV file
RAW_recipes_without_one_cuisine.to_csv('RAW_recipes_without_one_cuisine.csv', index=False)

# Verify the file creation by printing a message
print("RAW_recipes_without_one_cuisine DataFrame has been written to RAW_recipes_without_one_cuisine.csv")

RAW_recipes_without_one_cuisine DataFrame has been written to RAW_recipes_without_one_cuisine.csv
